### ISSUE TITLE
More issues with pcntl_fork

### ISSUE THREAD

**Issue by jacqueswaller** [mongodb/mongo-php-driver#527]
We ran into an issue that is similar to #174, except it affects php driver 1.2.x and doesn't require any sort of mongo activity in the child processes.

I have the following test script
```
<?php
    $manager = new MongoDB\Driver\Manager("localhost, [], ["typeMap" => ["root" => "array", "document" => "array"]]);
    $query = new MongoDB\Driver\Query([], []);
    $results = $manager->executeQuery("db.collection", $query);

    unset($manager); // unset manager to avoid conflict with child process (not necessary to repro; just here to show that doing so doesn't fix the problem)
    
    $parentPID = getmypid();
    
    $pid = pcntl_fork();
    
    if(getmypid() != $parentPID)
    {
        // Literally do nothing; just the process terminating causes the original mongo driver stream to do cleanup
        exit;
    }
    
    $pid = pcntl_waitpid(-1, $status); // wait for child
    
    $manager = new MongoDB\Driver\Manager("localhost", [], ["typeMap" => ["root" => "array", "document" => "array"]]);
    $results = $manager->executeQuery("db.collection", $query);
    
    exit;
?>
```

Upon executing the second query, you'll get some kind of ConnectionTimeoutException; sometimes the message is "Stream is closed", sometimes you get "Broken pipe (32)". This indicates to me that the act of the child process terminating does some kind of connection cleanup which can then permanently affect the parent process.

Just like in #174, you can create the 2nd manager with some different options and things will work just fine so we've written some silly workarounds for the one place we make use of multiple processes. Although it stands to reason that the child process cleanup shouldn't affect the parent, might I suggest that the connection persistence stuff should always create new if the existing connection is already closed?

If relevant, the platform I ran the test script on is ubuntu 14.04.5 x86_64, mongo version 3.0.12 and php 5.5.9 although our affected production code was on a variety of machines with different versions of stuff.

**Comment by jacqueswaller** [comment]
Sad times :( our workaround of "create a new Manager with different settings every time we have to fork a child" has led to #479; after awhile we get `ConnectionTimeoutException - No suitable servers found`.

**Comment by jmikola** [comment]
Thanks for the reproduction script. I created two issues that should address this: [PHPC-912](https://jira.mongodb.org/browse/PHPC-912) and [PHPC-913](https://jira.mongodb.org/browse/PHPC-913).

Can you elaborate on how the workaround of "creating a new Manager with different settings every time we have to fork a child" changes the original script and where the ConnectionTimeoutException is thrown (child or parent)?

**Comment by jacqueswaller** [comment]
You just have to modify the previous script to run in a loop making use of the "change the settings when creating the manager" workaround.
```
<?php
    $globalHackCounter = 30;

    while(true)
    {
        $manager = new MongoDB\Driver\Manager("localhost", ["connectTimeoutMS" => $globalHackCounter], ["typeMap" => ["root" => "array", "document" => "array"]]);
        $query = new MongoDB\Driver\Query([], []);
        $results = $manager->executeQuery("db.collection", $query);

        unset($manager); // unset manager to avoid conflict with child process (not necessary to repro; just here to show that doing so doesn't fix the problem)

        $parentPID = getmypid();

        $pid = pcntl_fork();

        if(getmypid() != $parentPID)
        {
            // Literally do nothing; just the process terminating causes the original mongo driver stream to do cleanup
            exit;
        }

        $pid = pcntl_waitpid(-1, $status); // wait for child
        $globalHackCounter += 1;
        print("$globalHackCounter\n");
    }
?>
```
By changing the connectTimeoutMS at each iteration, I avoid the immediate exception that you'd see with the original script. But just like in #479, I get `ConnectionTimeoutException - No suitable servers found` after 200-300 iterations.

**Comment by jacqueswaller** [comment]
Ok, on a hunch I tried without pcntl_fork() and the same issue occurs, same number of iterations, etc. So it's probably just some kind of maximum number of entries in the persistent hash table or something. Simplified repro script below:

```
<?php
    $globalHackCounter = 30;

    while(true)
    {
        $manager = new MongoDB\Driver\Manager("localhost", ["connectTimeoutMS" => $globalHackCounter], ["typeMap" => ["root" => "array", "document" => "array"]]);
        $query = new MongoDB\Driver\Query([], []);
        $results = $manager->executeQuery("db.collection", $query);

        unset($manager);

        $globalHackCounter += 1;
        print("$globalHackCounter\n");
    }
?>
```

Strictly speaking, if PHPC-912 and PHPC-913 are solved I won't personally have a use case for creating this many "unique" managers, but yeah.

**Comment by jmikola** [comment]
The above issues have been released in [1.2.6](https://github.com/mongodb/mongo-php-driver/releases/tag/1.2.6).
